File: python/rapidsmpf/rapidsmpf/tests/test_allgather_packed_data.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from typing import TYPE_CHECKING

from rapidsmpf.coll.allgather import AllGather
from rapidsmpf.memory.buffer_resource import BufferResource
from rapidsmpf.memory.packed_data import PackedData

if TYPE_CHECKING:
    import rmm

    from rapidsmpf.communicator.communicator import Communicator


def test_allgather_packed_data_roundtrip(
    comm: Communicator,
    device_mr: rmm.mr.CudaMemoryResource,
) -> None:
    br = BufferResource(device_mr)
    packed = PackedData.from_host_bytes(b"hello world", br)

    gather = AllGather(comm, op_id=1, br=br)
    gather.insert(0, packed)
    gather.insert_finished()

    result = gather.wait_and_extract(ordered=True)
    assert len(result) == 1
    assert result[0].to_host_bytes() == b"hello world"

File: cpp/tests/test_allgather_packed_data.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <chrono>

#include <gtest/gtest.h>

#include <rmm/mr/cuda_memory_resource.hpp>

#include <rapidsmpf/coll/allgather.hpp>
#include <rapidsmpf/communicator/single.hpp>
#include <rapidsmpf/memory/packed_data.hpp>

#include "environment.hpp"
#include "utils.hpp"

extern Environment* GlobalEnvironment;

TEST(AllGatherPackedData, basic_roundtrip) {
    auto const& comm = GlobalEnvironment->comm_;
    auto br = rapidsmpf::BufferResource::create(rmm::mr::cuda_memory_resource{});
    auto stream = br->stream_pool().get_stream();

    rapidsmpf::coll::AllGather allgather{comm, 0, br.get()};

    auto packed_data = generate_packed_data(4, 0, stream, *br);
    allgather.insert(0, std::move(packed_data));
    allgather.insert_finished();

    auto results = allgather.wait_and_extract(
        rapidsmpf::coll::AllGather::Ordered::NO, std::chrono::seconds{30}
    );
    EXPECT_FALSE(results.empty());
}
